fabric-data-attachment-api-v1/src/main/java/net/fabricmc/fabric/api/attachment/v1/AttachmentType.java

```diff
@@ -0,0 +1,211 @@
+ * This is done using the various {@code for*} factory methods:
+ * <ul>
+ *     <li>{@link #forBlockEntity} for block entities.</li>
+ *     <li>{@link #forChunk} for chunks.</li>
+ *     <li>{@link #forEntity} for entities.</li>
+ *     <li>{@link #forWorld} for worlds.</li>
+ * </ul>
+ *
+ * @param <A> Type of the attached object.
+ * @param <T> Type of the host of this attachment.
+ */
+@ApiStatus.NonExtendable
+public interface AttachmentType<A, T> {
+	/**
+	 * Create a new block entity attachment type with a given identifier, an attachment class, and optionally a serializer.
+	 *
+	 * <p>Serialization will only be used on the logical server.
+	 * Make sure to call {@link BlockEntity#markDirty()} after a change to the data attachments to ensure that they get persisted.
+	 *
+	 * @throws IllegalArgumentException If an attachment type with the given identifier already exists for block entities.
+	 */
+	static <A> AttachmentType<A, BlockEntity> forBlockEntity(Identifier identifier, Class<A> attachmentClass, @Nullable Serializer<A, ? super BlockEntity> serializer) {
+		return AttachmentTypeImpl.create(identifier, attachmentClass, BlockEntity.class, serializer);
+	}
+
+	/**
+	 * Create a new block entity attachment type with a given identifier, an attachment class, and optionally a serializer.
+	 *
+	 * <p>Serialization will only be used on the logical server.
+	 * Make sure to call {@link WorldChunk#setNeedsSaving} after a change to the data attachments to ensure that they get persisted.
+	 *
+	 * @throws IllegalArgumentException If an attachment type with the given identifier already exists for block entities.
+	 */
+	static <A> AttachmentType<A, WorldChunk> forChunk(Identifier identifier, Class<A> attachmentClass, @Nullable Serializer<A, ? super WorldChunk> serializer) {
+		return AttachmentTypeImpl.create(identifier, attachmentClass, WorldChunk.class, serializer);
+	}
+
+	/**
+	 * Create a new block entity attachment type with a given identifier, an attachment class, and optionally a serializer.
+	 *
+	 * <p>Serialization will only be used on the logical server.
+	 * Entity data is always persisted, there is no need to call any method after modifying the attachments.
+	 *
+	 * <p>Note that data is not transferred automatically in the following cases, and requires special handling if desired:
+	 * <ul>
+	 *     <li>On the (logical) client and server: when a player dies and respawns.
+	 *     {@code ServerPlayerEvents.COPY_FROM} can be used to copy the data on the server.</li>
+	 *     <li>On the client: when a player changes dimensions.</li>
+	 *     <li>On the client and server: when a non-player entity changes dimensions.</li>
+	 * </ul>
+	 *
+	 * @throws IllegalArgumentException If an attachment type with the given identifier already exists for block entities.
+	 */
+	static <A> AttachmentType<A, Entity> forEntity(Identifier identifier, Class<A> attachmentClass, @Nullable Serializer<A, ? super Entity> serializer) {
```

Comment:
A general concern is that currently the attachments are not synced to the client. While it is debatable whether block entity, chunk and world attachments should be synced, I think it's pretty clear that syncing entity attachments is very useful.

Thoughts on adding optional entity attachment syncing? It might make implementation quite a bit messier (see https://github.com/MinecraftForge/MinecraftForge/pull/8706), but it will definitely make the API more useful I think. Block entity might also not be too hard to add.

Reply:
I decided that it is too hard for now, but it can be added later if desired. See https://github.com/FabricMC/fabric/pull/2172#issuecomment-1256576448.